blockchain/v2/reactor_test.go

```diff
@@ -1,22 +1,552 @@
+				{"Receive", "P2", bcStatusResponseMessage{Height: 15}},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P1", bcBlockResponseMessage{refStore.LoadBlock(11)}},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(12)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P1", bcBlockResponseMessage{refStore.LoadBlock(13)}},
+				{"Process", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(14)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(15)}},
+				{"Process", "", nil},
+			},
+		},
+		{
+			name:   "termination on max peer height - two peers, noBlock error",
+			params: params,
+			msgs: []testEvent{
+				{"AddPeer", "P1", nil},
+				{"AddPeer", "P2", nil},
+				{"Receive", "P1", bcStatusResponseMessage{Height: 13}},
+				{"Receive", "P2", bcStatusResponseMessage{Height: 15}},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P1", bcNoBlockResponseMessage{11}},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(12)}},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(11)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(13)}},
+				{"Process", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(14)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(15)}},
+				{"Process", "", nil},
+			},
+		},
+		{
+			name:   "termination on max peer height - two peers, remove one peer",
+			params: params,
+			msgs: []testEvent{
+				{"AddPeer", "P1", nil},
+				{"AddPeer", "P2", nil},
+				{"Receive", "P1", bcStatusResponseMessage{Height: 13}},
+				{"Receive", "P2", bcStatusResponseMessage{Height: 15}},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"RemovePeer", "P1", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(12)}},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(11)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(13)}},
+				{"Process", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(14)}},
+				{"Process", "", nil},
+				{"BlockReq", "", nil},
+				{"Receive", "P2", bcBlockResponseMessage{refStore.LoadBlock(15)}},
+				{"Process", "", nil},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			reactor := newTestReactor(params)
+			reactor.Start()
+			mockSwitch := &mockSwitchIo{switchedToConsensus: false}
+			reactor.io = mockSwitch
+			// time for go routines to start
+			time.Sleep(time.Millisecond)
```

Comment:
This should not be necessary, events should buffer until the goroutine picks them up.